From 8a3038b6c7d22c09b66597329d6e14b30d294f16 Mon Sep 17 00:00:00 2001
From: Martin Braun <martin.braun@ettus.com>
Date: Tue, 22 Aug 2017 18:13:37 -0700
Subject: [PATCH 2/3] C API: Dboard EEPROM revision error handling fix

For invalid dboard revisions stored in the EEPROM, provide a better
error message.
---
 host/include/uhd/usrp/dboard_eeprom.h |  6 +++++-
 host/lib/usrp/dboard_eeprom_c.cpp     | 19 ++++++++++++++++++-
 2 files changed, 23 insertions(+), 2 deletions(-)

diff --git a/host/include/uhd/usrp/dboard_eeprom.h b/host/include/uhd/usrp/dboard_eeprom.h
index 6980de0ce..f3b9c4d9c 100644
--- a/host/include/uhd/usrp/dboard_eeprom.h
+++ b/host/include/uhd/usrp/dboard_eeprom.h
@@ -84,7 +84,11 @@ UHD_API uhd_error uhd_dboard_eeprom_set_serial(
     const char* serial
 );
 
-//! Get the daughterboard's revision (not always present)
+/*! Get the daughterboard's revision
+ *
+ * The revision doesn't always have to be present, in which case this function
+ * will return an error.
+ */
 UHD_API uhd_error uhd_dboard_eeprom_get_revision(
     uhd_dboard_eeprom_handle h,
     int* revision_out
diff --git a/host/lib/usrp/dboard_eeprom_c.cpp b/host/lib/usrp/dboard_eeprom_c.cpp
index e3ef4933f..9bfa48b66 100644
--- a/host/lib/usrp/dboard_eeprom_c.cpp
+++ b/host/lib/usrp/dboard_eeprom_c.cpp
@@ -19,6 +19,7 @@
 #include <uhd/error.h>
 
 #include <boost/lexical_cast.hpp>
+#include <boost/format.hpp>
 
 #include <string.h>
 
@@ -79,12 +80,28 @@ uhd_error uhd_dboard_eeprom_set_serial(
     )
 }
 
+//! Convert a string into an int. If that doesn't work, craft our own exception
+// instead of using the Boost exception. We need to put this separate from the
+// caller function because of macro expansion.
+int _convert_rev_with_exception(const std::string &rev_str)
+{
+    try {
+        return boost::lexical_cast<int>(rev_str);
+    } catch (const boost::bad_lexical_cast &) {
+        throw uhd::lookup_error(str(
+            boost::format("Error retrieving revision from string `%s`")
+            % rev_str
+        ));
+    }
+}
+
 uhd_error uhd_dboard_eeprom_get_revision(
     uhd_dboard_eeprom_handle h,
     int* revision_out
 ){
     UHD_SAFE_C_SAVE_ERROR(h,
-        *revision_out = boost::lexical_cast<int>(h->dboard_eeprom_cpp.revision);
+        *revision_out = \
+            _convert_rev_with_exception(h->dboard_eeprom_cpp.revision);
     )
 }
 
-- 
2.11.0

